### PR TITLE
Add support for structured sparsity (block structure & n:m structure)

### PR DESCRIPTION
**Pull request by u-simha**
## What
This PR adds support for structured sparsity, specifically block structured & n:m structured. Certain hardware can accelerate sparsity further by skipping blocks of 0 when loading the weights into memory. 

## How
- Move the mask computation out of `prune.py` and abstracted cleanly into `scheme.py` where each pruning scheme computes the mask based on the sparsity
- Add logic to zero out the lowest magnitude blocks based on block size or n:m ratio
- Minor refactoring to bubble up `BlockSizeMismatch` error to be shared across pruning & quantization

## Testing
- Update magnitude pruner tests to check for block sparsity and n:m sparsity
- Update config / spec tests to check validations for block size and n:m ratio

### CODE REVIEW

File: src/coreai_opt/pruning/spec/errors.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2026 Apple Inc.
```
Comment (u-simha) [inline src/coreai_opt/pruning/spec/errors.py]: Directly use `_utils.errors._BlockSizeMismatchError` instead of creating copies of errors.py in pruning & quantization

File: src/coreai_opt/_utils/errors.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2026 Apple Inc.
```
Comment (guru-desh) [inline src/coreai_opt/_utils/errors.py:1]: Will you update the pruning tutorial to show structured sparsity in this PR?

File: src/coreai_opt/pruning/spec/errors.py (hunk 1 above)
Comment (u-simha) [reply src/coreai_opt/pruning/spec/errors.py]: Fixed in 4a5a166 — removed both errors.py shims (pruning & quantization) and updated all call sites to import `_BlockSizeMismatchError` directly from `coreai_opt._utils.errors`.

File: src/coreai_opt/_utils/errors.py (hunk 2 above)
Comment (u-simha) [reply src/coreai_opt/_utils/errors.py:1]: I don't plan to update the tutorial, but I did plan to update the doc page, specifically the config page to highlight the options. Let me add that onto this PR itself.

File: src/coreai_opt/pruning/spec/scheme.py [hunk 3]
```diff
@@ -51,6 +65,32 @@ def _serialize_model(self) -> dict[str, Any]:
...
+            0 = prune).
+        """
+        if sparsity == 0.0:
+            return torch.ones_like(weight)
+        if sparsity >= 1.0:
+            return torch.zeros_like(weight)
+        return self._compute_mask(weight, sparsity)
+
+    @abstractmethod
+    def _compute_mask(self, weight: torch.Tensor, sparsity: float) -> torch.Tensor:
```
Comment (crowbat) [inline src/coreai_opt/pruning/spec/scheme.py:86]: Given that `PruningScheme` is a public class meant for users to inherit from, I think we should have any abstractmethods users must override be public as well. Otherwise the private naming suggests it's a method we can choose to freely modify, but would cause any user defined code to break if we do

File: src/coreai_opt/pruning/spec/scheme.py [hunk 4]
```diff
@@ -62,17 +102,170 @@ class Unstructured(PruningScheme):
...
+
+@PruningScheme.register("n_m_structured")
+class NMStructured(PruningScheme):
+    """N:M structured pruning scheme.
+
+    Zeroes exactly ``n`` smallest-magnitude elements out of every contiguous
+    group of ``m`` elements along ``axis`` — a hardware-friendly sparsity
+    pattern with a fixed sparsity ratio of ``n / m``.
+
+    Unlike other schemes, the achieved sparsity is fixed by construction and
```
Comment (crowbat) [inline src/coreai_opt/pruning/spec/scheme.py:217]: Instead of just ignoring target_sparsity completely which could be misleading (it is a public object which user logic may depend on), could we instead have it be derived from doing n/m?

And then two things could come out as a result:
1) No need to override the existing `compute_mask` method (it can follow the early exit logic of sparsity = 0 or 1)
2) When defining the abstract `_compute_mask` method, instead of always using strict n/m and ignoring sparsity, it can still use sparsity to determine the number of elements within each block to zero out. This would allow N:M structured to also be trainable via a schedule. We could have a statement like `num_prune = self.n if sparsity >= self.n / self.m else math.floor(sparsity *
  self.m)` to protect against numerical inaccuracies in doing `sparsity * self.m`, to guarantee that we always end up with at least n elements zeroed out at the end of training.

We could make target_sparsity a `float | None` in `PruningSpec` where it is required if the pruning scheme is not NMStructured, and raise an error if the user creates a spec with both target_sparsity provided as well as pruning_scheme = NMStructured.

File: src/coreai_opt/pruning/spec/scheme.py [hunk 5]
```diff
@@ -62,17 +102,170 @@ class Unstructured(PruningScheme):
...
+        if num_prune == 0:
+            return torch.ones_like(weight)
+        if num_prune >= num_blocks:
+            return torch.zeros_like(weight)
+
+        # axis is now at position 0; all other dims keep their relative order.
+        moved = torch.movedim(weight, self.axis, 0)
+        other_dims = moved.shape[1:]
+        grouped = moved.view(num_blocks, self.block_size, *other_dims)
+        block_norms = grouped.pow(2).sum(dim=tuple(range(1, grouped.ndim))).sqrt()
```
Comment (crowbat) [inline src/coreai_opt/pruning/spec/scheme.py:194]: Nit, this could be rewritten as `torch.linalg.vector_norm(grouped, dim=...)`
